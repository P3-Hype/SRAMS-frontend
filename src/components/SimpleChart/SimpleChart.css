.simple-chart-canvas {
    width: 100%;
    height: 100%;
}

.simple-chart {
    width: 100%;
    height: 100%;
}