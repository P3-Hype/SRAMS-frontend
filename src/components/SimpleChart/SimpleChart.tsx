import "./SimpleChart.css";
import { useEffect, useRef, useState } from "react";

interface SimpleChartProps {
    readonly title: string
    readonly data: number[]
    readonly lineColor?:string
    readonly dataOffset?: number
    readonly width?: number
    readonly height?: number
}

export function SimpleChart(props: SimpleChartProps) {
    const [canvasWidth, setCanvasWidth] = useState(0)
    const [canvasHeight, setCanvasHeight] = useState(0)
    const canvasRef = useRef<HTMLCanvasElement>(null);
    const wrapperRef = useRef<HTMLDivElement>(null);
    const lineColor = props.lineColor ?? '#000000';

    useEffect(() => {
        if (!canvasRef.current) return;
        const data = props.data;
        const dataMax = Math.max(...data) - (props.dataOffset ?? 0);
        //const dataMin = Math.min(...data)
        const ctx = canvasRef.current.getContext("2d");
        if (!ctx) return;
        console.log(ctx);
        
        ctx.clearRect(0,0,canvasRef.current.width, canvasRef.current.height);
        ctx.strokeStyle = lineColor;
        ctx.lineWidth = 2;
        ctx.beginPath();
        ctx.moveTo(0, (1-(data[0] / dataMax)) * canvasHeight);
        for (let i = 0; i < data.length; i++) {
            const y = data[i] - (props.dataOffset ?? 0);
            const mappedY =  (1-(y / dataMax)) * canvasHeight;
            const mappedX = (i / (data.length-1)) * canvasWidth;
            ctx.lineTo(mappedX, mappedY);
        }
        ctx.stroke();
        ctx.strokeStyle = "#242424";
        ctx.lineWidth = 8;
        ctx.fillStyle = lineColor;
        ctx.font = "16pt Arial";
        ctx.strokeText(props.title, 0, canvasHeight);
        ctx.fillText(props.title, 0, canvasHeight);
        
    }, [canvasHeight, canvasWidth, lineColor, props.data, props.dataOffset, props.title]);

    useEffect(() => {
        if (wrapperRef.current == null) return
        const width = wrapperRef.current?.clientWidth;
        const height = wrapperRef.current?.clientHeight;
        setCanvasWidth(width)
        setCanvasHeight(height)
    }, [wrapperRef]);
    
    return (
        <div className="simple-chart" ref={wrapperRef} style={{
            width: (props.width) ? props.width + "px" : "100%",
            height: (props.height) ? props.height + "px" : "100%"
        }}>
            <canvas className="simple-chart-canvas" ref={canvasRef} width={canvasWidth} height={canvasHeight} />
        </div>
    );
}

export default SimpleChart;