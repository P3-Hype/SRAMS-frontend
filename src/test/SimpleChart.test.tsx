import { expect, it, describe } from 'vitest'
import renderer from 'react-test-renderer'
import SimpleChart from '../components/SimpleChart/SimpleChart'
//import { render, screen } from '@testing-library/react'

const data = [59,65,555,674,43,554,426,4984,513,584]

describe("SimpleChart tests", () => {
    it("renders correctly", () => {
        const component = renderer.create(
            <SimpleChart 
            title='test chart' 
            data={data}
            width={300}
            height={100}
            lineColor='#ff0000'/>
        );
        expect(component.toJSON()).toMatchSnapshot();
    })

    // it("renders with empty dataset", () => {
    //     render(
    //         <SimpleChart 
    //         title='test chart' 
    //         data={[]}
    //         width={300}
    //         height={100}
    //         lineColor='#ff0000'/>
    //     );
    //     const simpleChartElement = screen.getByTestId("simple-chart");
    //     expect(simpleChartElement).toBeInTheDocument();
    // })

    // it("renders with only one entry in dataset", () => {
    //     render(
    //         <SimpleChart 
    //         title='test chart' 
    //         data={[1]}
    //         width={300}
    //         height={100}
    //         lineColor='#ff0000'/>
    //     );
    //     const simpleChartElement = screen.getByTestId("simple-chart");
    //     expect(simpleChartElement).toBeInTheDocument();
    // })
})