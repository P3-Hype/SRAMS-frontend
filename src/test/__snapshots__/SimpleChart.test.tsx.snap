// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SimpleChart > renders correctly 1`] = `
<div
  className="simple-chart"
  style={
    {
      "height": "100px",
      "width": "300px",
    }
  }
>
  <canvas
    className="simple-chart-canvas"
    height={0}
    width={0}
  />
</div>
`;

exports[`SimpleChart > renders with only one entry in dataset 1`] = `
<div
  className="simple-chart"
  style={
    {
      "height": "100px",
      "width": "300px",
    }
  }
>
  <canvas
    className="simple-chart-canvas"
    height={0}
    width={0}
  />
</div>
`;

exports[`SimpleChart tests > renders correctly 1`] = `
<div
  className="simple-chart"
  style={
    {
      "height": "100px",
      "width": "300px",
    }
  }
>
  <canvas
    className="simple-chart-canvas"
    height={0}
    width={0}
  />
</div>
`;
